### ISSUE TITLE
Running as non-root user

### ISSUE THREAD

**Issue by TG-March** [Lynx-Shortener/Lynx#153]
I'd like to run node as a non-root user in my container. I'd actually love for this to be an option in the official release!

I'm not very familiar with node development or developing docker containers, so hoping you could point me in the right direction.

I have modified the `Dockerfile` as below.

```
# Define non-root user
ARG USER_NAME=node
ARG USER_ID=1003
ARG GROUP_ID=1003

FROM node:20 AS BUILD_IMAGE

ARG USER_NAME
ARG USER_ID
ARG GROUP_ID

ENV USERNAME=$USER_NAME
ENV USER_UID=$USER_ID
ENV USER_GID=$GROUP_ID

RUN echo ${USERNAME}
RUN echo ${USER_UID}
RUN echo ${USER_GID}


# Create the user
#RUN groupadd --gid $USER_GID $USERNAME \
#    && useradd --uid $USER_UID --gid $USER_GID -m $USERNAME
    # [Optional] Add sudo support. Omit if you don't need to install software after connecting.
    #&& apt-get update \
    #&& apt-get install -y sudo \
    #&& echo $USERNAME ALL=\(root\) NOPASSWD:ALL > /etc/sudoers.d/$USERNAME \
    #&& chmod 0440 /etc/sudoers.d/$USERNAME	

RUN groupmod --gid $USER_GID $USERNAME \
    && usermod --uid $USER_UID --gid $USER_GID $USERNAME \
    && chown -R $USER_UID:$USER_GID /home/$USERNAME
	
ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable

RUN mkdir /home/$USERNAME/app
RUN mkdir /home/$USERNAME/app/node_modules
RUN mkdir /home/$USERNAME/app/node_modules/.cache
RUN chown -R $USER_UID:$USER_GID /home/$USERNAME/app

WORKDIR /home/$USERNAME/app

# Copy required files for packages 

COPY --chown=$USER_UID:$USER_GID ["./frontend/","/home/node/app/frontend/"]

COPY --chown=$USER_UID:$USER_GID ["./src/package.json","./src/pnpm-lock.yaml", "/home/node/app/src/"]

COPY --chown=$USER_UID:$USER_GID ["VERSION", "./"]

# Install dependencies for backend

RUN pnpm install --frozen-lockfile --recursive

COPY --chown=$USER_UID:$USER_GID src src

# Install dependencies for frontend and build it

WORKDIR /home/$USERNAME/app/frontend
RUN NODE_ENV=production
RUN pnpm build

FROM gcr.io/distroless/nodejs18-debian11

ARG USER_NAME
ARG USER_ID
ARG GROUP_ID

ENV USERNAME=$USER_NAME
ENV USER_UID=$USER_ID
ENV USER_GID=$GROUP_ID

## From the builder image above, start node from a distroless image

WORKDIR /home/$USERNAME/app

COPY --chown=$USER_UID:$USER_GID --from=BUILD_IMAGE /home/$USERNAME/app/dist ./dist
COPY --chown=$USER_UID:$USER_GID --from=BUILD_IMAGE /home/$USERNAME/app/src ./src
COPY --chown=$USER_UID:$USER_GID --from=BUILD_IMAGE /home/$USERNAME/app/VERSION ./VERSION

EXPOSE 3000

USER $USER_UID:$USER_GID

CMD ['/home/node/app/src/index.js']


```

Here is the error I'm getting in the logs:

```
Node.js v18.20.3
 node:internal/modules/cjs/loader:1143
   throw err;
   ^

 Error: Cannot find module '/bin/sh'
     at Module._resolveFilename (node:internal/modules/cjs/loader:1140:15)
     at Module._load (node:internal/modules/cjs/loader:981:27)
     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:128:12)
     at node:internal/main/run_main_module:28:49 {
   code: 'MODULE_NOT_FOUND',
   requireStack: []
 }
```

**Comment by TG-March** [comment]
I think it's path related because if I change the final CMD to 

`CMD ["node", "/home/node/app/src/index.js"]`

then I think I'm a little closer...

```
Error: Cannot find module '/home/node/app/node'
     at Module._resolveFilename (node:internal/modules/cjs/loader:1140:15)
     at Module._load (node:internal/modules/cjs/loader:981:27)
     at Function.executeUserEntryPoint [as runMain] (node:internal/modules/run_main:128:12)
     at node:internal/main/run_main_module:28:49 {
   code: 'MODULE_NOT_FOUND',
   requireStack: []
 }
```

**Comment by JackBailey** [comment]
Hi, thank you for your initial work on this. It looks like the distroless image I've used for this does actually have a `:nonroot` tag. I'll try to play around with it, but my knowledge on permissions isn't great. Looks like it uses `65532`, and working dir `/home/nonroot/`

https://stackoverflow.com/a/74811728
https://github.com/GoogleContainerTools/distroless?tab=readme-ov-file#debian-12

**Comment by TG-March** [comment]
Thanks, I appreciate your quick reply.

Below is my Dockerfile with your suggestion.
This has been much more successful and the app is now running, however unfortunately still falls over early on.

```
FROM node:20 AS BUILD_IMAGE

ENV PNPM_HOME="/pnpm"
ENV PATH="$PNPM_HOME:$PATH"
RUN corepack enable

WORKDIR /app

# Copy required files for packages 

COPY ["./frontend/","/app/frontend/"]

COPY ["./src/package.json","./src/pnpm-lock.yaml", "/app/src/"]

COPY ["VERSION", "./"]

# Install dependencies for backend

RUN pnpm install --frozen-lockfile --recursive

COPY src src

# Install dependencies for frontend and build it

WORKDIR /app/frontend
RUN NODE_ENV=production
RUN pnpm build

FROM gcr.io/distroless/nodejs18-debian11:nonroot

## From the builder image above, start node from a distroless image

WORKDIR /app

COPY --chown=65532:65532 --from=BUILD_IMAGE /app/dist ./dist
COPY --chown=65532:65532 --from=BUILD_IMAGE /app/src ./src
COPY --chown=65532:65532 --from=BUILD_IMAGE /app/VERSION ./VERSION

EXPOSE 3000

CMD ["/app/src/index.js"]
```

On the host I have run chown -R 65532:65532 on the lynx/ directory and then chown -R some_user:some_user on the mongodb subdirectory. Mongodb is running without issue.

```
drwxrwxr-x 6        65532 nonroot      4096 Jun 11 18:05 ./
drwxrwx--- 5 some_user    some_user    4096 Jun  7 18:05 ../
drwxr-xr-x 4        65532 nonroot      4096 Jun  7 17:52 app/
drwxr-xr-x 2        65532 nonroot      4096 Jun 11 18:05 backups/
drwxrwxr-x 4 some_user    some_user    4096 Jun  1 18:37 mongodb/
drwxr-xr-x 3        65532 nonroot      4096 Jun 11 17:46 tmp/
```

My new log file shows:

```
Error: EACCES: permission denied, mkdir '../backups'
     at Object.mkdirSync (node:fs:1391:3)
     at module.exports (/app/src/modules/backup.js:7:12)
     at process.processTicksAndRejections (node:internal/process/task_queues:95:5) {
   errno: -13,
   syscall: 'mkdir',
   code: 'EACCES',
   path: '../backups'
 }
```

Interestingly if I delete lynx/backups on the host, it is recreated with root:root ownership on lynx startup, however the error persists. 

My docker compose mounts the volumes as follows

```
        volumes:
            - /home/some_user/appdata/lynx/backups:/app/backups
            - /home/some_user/appdata/lynx/tmp:/app/tmp
```

so it's strange that in backups.js, `fs.existsSync("../backups")` returns false and tries to create the directory.

Creating the user nonroot with group nonroot with uid:gid 65532:65532 on the host has not helped.

I have confirmed the process is run with uid 65532 as well:

```
65532 | 1250595 | 1250575 | 84 | 18:05 | ? | 00:00:00 | /nodejs/bin/node /app/src/index.js
```

I have tried with and without `user: 65532:65532` in the `docker-compose.yml` but makes no difference.

**Comment by TG-March** [comment]
Changing all of the relative paths in backup.js from `../backups` to `/app/backups` has allowed the app to startup!

I don't have enough time to test anything else tonight, but will report back soon!

**Comment by JackBailey** [comment]
Closed due to #155.
